module/zfs/mmp.c

```diff
@@ -258,27 +235,40 @@ mmp_random_leaf_impl(vdev_t *vd, int *fail_mask)
  * MMP_FAIL_NOT_WRITABLE
  * If set, one or more vdevs are not writeable.  The children of those vdevs
  * were not examined.
- *
- * Assuming in_vd points to a tree, a random subtree will be chosen to start.
- * That subtree, and successive ones, will be walked until a usable leaf has
- * been found, or all subtrees have been examined (except that the children of
- * un-writeable vdevs are not examined).
- *
- * If the leaf vdevs in the tree are healthy, the distribution of returned leaf
- * vdevs will be even.  If there are unhealthy leaves, the following leaves
- * (child_index % index_children) will be chosen more often.
  */
 
 static int
-mmp_random_leaf(vdev_t *in_vd, vdev_t **out_vd)
+mmp_next_leaf(spa_t *spa, vdev_t **current_leaf)
```

Comment:
Why is this second parameter `vdev_t **current_leaf` needed? Can't we already access the current leaf as `spa->mmp_last_leaf`?

```diff
diff --git a/module/zfs/mmp.c b/module/zfs/mmp.c
index fc81f3e..926b7de 100644
--- a/module/zfs/mmp.c
+++ b/module/zfs/mmp.c
@@ -238,18 +238,17 @@ mmp_list_next_wrap(list_t *list, vdev_t *leaf)
  */
 
 static int
-mmp_next_leaf(spa_t *spa, vdev_t **current_leaf)
+mmp_next_leaf(spa_t *spa)
 {
        vdev_t *leaf;
        boolean_t found = B_FALSE;
        int fail_mask = 0;
 
-       ASSERT(current_leaf != NULL);
        ASSERT(list_link_active(&spa->spa_leaf_list.list_head) == B_TRUE);
        ASSERT(!list_is_empty(&spa->spa_leaf_list));
 
-       leaf = mmp_list_next_wrap(&spa->spa_leaf_list, *current_leaf);
-       *current_leaf = leaf;
+       leaf = mmp_list_next_wrap(&spa->spa_leaf_list, spa->mmp_last_leaf);
+       spa->mmp_last_leaf = leaf;
 
        do {
                ASSERT(leaf != NULL);
@@ -262,11 +261,11 @@ mmp_next_leaf(spa_t *spa, vdev_t **current_leaf)
                        fail_mask |= MMP_FAIL_WRITE_PENDING;
                } else {
                        found = B_TRUE;
-                       *current_leaf = leaf;
+                       spa->mmp_last_leaf = leaf;
                }
 
                leaf = mmp_list_next_wrap(&spa->spa_leaf_list, leaf);
-       } while (found == B_FALSE && leaf != *current_leaf);
+       } while (found == B_FALSE && leaf != spa->mmp_last_leaf);
 
        return (found == B_TRUE ? 0 : fail_mask);
 }
@@ -390,7 +389,7 @@ mmp_write_uberblock(spa_t *spa)
 
        mutex_enter(&mmp->mmp_io_lock);
 
-       error = mmp_next_leaf(spa, &spa->mmp_last_leaf);
+       error = mmp_next_leaf(spa);
 
        /*
         * spa_mmp_history has two types of entries:
```

Reply:
You're right, that would be better.

Reply:
Rather than store the last leaf in the spa what if we instead rotated the leafs in the list.  Always removing from the head and reinserting at the tail.  Practically this isn't any more expensive, and it allows us to return the `mmp_leaf` to use next.  Something like this (untested):

```c
static int
mmp_next_leaf(spa_t *spa, vdev_t **mmp_leaf)
{
        int fail_mask = 0;

        ASSERT(!list_is_empty(&spa->spa_leaf_list));
        vdev_t *first_leaf = list_head(&spa->spa_leaf_list);

        do {
                vdev_t *leaf = list_remove_head(&spa->spa_leaf_list);
                list_insert_tail(&spa->spa_leaf_list, leaf);

                if (!vdev_writeable(leaf)) {
                        fail_mask |= MMP_FAIL_NOT_WRITABLE;
                } else if (leaf->vdev_mmp_pending != 0) {
                        fail_mask |= MMP_FAIL_WRITE_PENDING;
                } else {
                        *mmp_leaf = leaf;
                        return (0);
                }
        } while (list_head(&spa->spa_leaf_list) != start_leaf);

        return (fail_mask);
}
```